### PR TITLE
Feature json expression

### PR DESCRIPTION
**Pull request by Deecellar**
Added feature following https://quire.io/w/Monofoxe/11/Add_support_for_m...?board=Monofoxe&filter=all 
Gray Foxes helped me in this one

- Added Naive implementation on tokenizer/parser
- PEMDAS is followed (even tho technically the E does not exist =D)

### CODE REVIEW

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupImporter.cs [hunk 1]
```diff
@@ -131,40 +131,12 @@ private void ImportTextures(string dirPath, string dirName, SpriteGroupData grou
 
 						int xOffset, yOffset;
 
-						// Hey, look, switch is being useful for once! :000
-
+						SpriteGroupMathParser parser = new SpriteGroupMathParser(spr);
```
Comment (Martenfur) [inline Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupImporter.cs]: Use var.

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+		}
+		/// <summary>
+		/// This recognize our terminal elements, like number and constants, also reduce parenthesis expresions
+		/// </summary>
+		/// <returns>Terminal element</returns>
+		Node ParseLeaf()
+		{
+			switch (_tokenizer.CurrentToken)
+			{
+				case Token.Number:
```
Comment (Martenfur) [inline Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs:175]: This is barely readable. Add braces to cases, or better yet - get rid of the switch.

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+					return new NumberNode(0);
+				case Token.ConstantCenter:
+					_tokenizer.NextToken();
+					if (_yAxis)
+					{
+						return new NumberNode(_rawSprite.RawTexture.Height / _rawSprite.FramesV / 2);
+					}
+					return new NumberNode(_rawSprite.RawTexture.Width / _rawSprite.FramesH / 2);
+				default:
+					throw new Exception($"Unexpect token: {_tokenizer.CurrentToken}"); // TODO: Make a SyntaxException or something similar
```
Comment (Martenfur) [inline Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs]: Forgot the TODO. : - )

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+	/// <summary>
+	/// This class parse the expresion and returns the number that corresponds to it
+	/// Might be slow since it uses recursion to parse the expresion
+	/// Not benchmarked yet
+	/// </summary>
+	//  Based on the code by Brad Robinson in this Medium Article https://medium.com/@toptensoftware/writing-a-simple-math-expression-engine-in-c-d414de18d4ce
+	public class SpriteGroupMathParser
+	{
+
+		public SpriteGroupMathParser(RawSprite sprite)
```
Comment (Martenfur) [inline Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs]: Make this class static, there is no reason for it to be instantiated.

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+		{
+			_rawSprite = sprite; // Maybe this should not be here but hey... Fox are sneaky
```
Comment (Martenfur) [inline Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs]: Yeah, it most definitely should not. I propose removing _rawSprite and _yAxis and passing constants directly into the Parse method.

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+					return new NumberNode(_rawSprite.RawTexture.Width / _rawSprite.FramesH / 2);
+				default:
+					throw new Exception($"Unexpect token: {_tokenizer.CurrentToken}"); // TODO: Make a SyntaxException or something similar
+			}
+		}
+
+	}
+	#region NodeClasses
+
+	abstract class Node
```
Comment (Martenfur) [inline Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs]: Why are all those in a single file?

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathTokenizer.cs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+	TextReader _textReader;
+	char _currentChar;
+	StringBuilder _builder;
+	string _text;
+	public SpriteGroupMathTokenizer(string text)
+	{
+	  _builder = new StringBuilder(text);
+	  _builder.Replace("center", "c").Replace("top", "t").Replace("bottom", "b").Replace("left", "l").Replace("right", "r");
+	  _text = _builder.ToString();
```
Comment (Martenfur) [inline Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathTokenizer.cs]: The string is way too longe, make it something like
```
_builder.Replace("center", "c")
  .Replace("top", "t")
  .Replace("bottom", "b")
```

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathTokenizer.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	  _textReader = new StringReader(_text);
+	  _builder.Clear();
+	  NextChar();
```
Comment (Martenfur) [inline Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathTokenizer.cs]: Those are indented with spaces, by the way. Check the rest to have the correct indents.

**Review by Martenfur — CHANGES_REQUESTED** [review]

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs (hunk 2 above)
Comment (Deecellar) [reply Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs:175]: Decided first form, not a fan of 
```csharp
if(_tokenizer.CurrentToken == Token.Number) 
{
	//...
}
```
with a lot of conditions

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs (hunk 3 above)
Comment (Deecellar) [reply Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs]: True =D not anymore

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs (hunk 4 above)
Comment (Deecellar) [reply Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs]: Done =3

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs (hunk 6 above)
Comment (Deecellar) [reply Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathParser.cs]: oh my, forgot to let refactor that when cleaning

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathTokenizer.cs (hunk 7 above)
Comment (Deecellar) [reply Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathTokenizer.cs]: Yuop

File: Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathTokenizer.cs (hunk 8 above)
Comment (Deecellar) [reply Monofoxe/Pipefoxe/SpriteGroup/SpriteGroupMathTokenizer.cs]: Codestyle ahhh

**Comment by Martenfur** [comment]
Since there were conflicts, I've just copied the code and pushed it locally. Gonna close this PR.

**Pull request closed without merging**